### PR TITLE
Removed the initiation of default outputs in assembler mode

### LINKED ISSUE #12230: [CLI] Do not select any outputs by default in assembler mode
#12074 added support for CLI output options (like `--bin` or `--ewasm`) in assembler mode. If no outputs are requested though, the compiler chooses the default ones to keep the behavior backwards-compatible. We should change this so that nothing is printed by default (this is how the compiler mode behaves) but it needs to be done in a breaking release.

### PR DESCRIPTION
**Pull request by mejsiej**
~Depends on #12376.~ Merged.
Closes #12230

### CODE REVIEW

File: Changelog.md
Comment (cameel) [inline Changelog.md]: How about this?
```suggestion
 * Commandline Interface: Assembler mode no longer enables all outputs by default.
```

File: test/cmdlineTests/linking_strict_assembly/args
Comment (cameel) [inline test/cmdlineTests/linking_strict_assembly/args]: In case of all the linker tests we can reduce this to `--bin`. All other outputs are not affected.
```suggestion
--strict-assembly --bin --libraries contract/test.sol:L=0x1234567890123456789012345678901234567890
```

File: test/cmdlineTests/strict_asm_optimizer_steps/args
Comment (cameel) [inline test/cmdlineTests/strict_asm_optimizer_steps/args]: Here only the IR output is relevant.
```suggestion
--strict-assembly --optimize --ir-optimized --yul-optimizations dhfoDgvulfnTUtnIf
```

File: test/cmdlineTests/yul_function_name_clashes/args
Comment (cameel) [inline test/cmdlineTests/yul_function_name_clashes/args]: The "name clashes" tests do not produce any ewasm code so ewasm output is superfluous..
```suggestion
--strict-assembly --debug-info none --bin --asm --ir-optimized
```

File: test/cmdlineTests/wasm_to_wasm_function_returning_multiple_values/args
Comment (cameel) [inline test/cmdlineTests/wasm_to_wasm_function_returning_multiple_values/args]: Currently you can only get output from either `--asm` or `--ewasm`, never both. `--asm` is generally not needed in any tests that use `ewasm` dialect and I'd leave only in cases where we actually want to verify that (e.g. in "debug info" tests). 
```suggestion
--yul --yul-dialect ewasm --machine ewasm --bin --ewasm --ewasm-ir --ir-optimized
```

File: test/cmdlineTests/yul_optimize_runs/args
Comment (cameel) [inline test/cmdlineTests/yul_optimize_runs/args]:
```suggestion
--yul --yul-dialect evm --optimize --bin --asm --ir-optimized --optimize-runs 10000
```

File: test/cmdlineTests/yul_verbatim/args
Comment (cameel) [inline test/cmdlineTests/yul_verbatim/args]:
```suggestion
--strict-assembly --bin --asm --ir-optimized
```

File: test/cmdlineTests/yul_verbatim_msize/args
Comment (cameel) [inline test/cmdlineTests/yul_verbatim_msize/args]:
```suggestion
--strict-assembly --bin --asm --ir-optimized --optimize
```

**Review by cameel — CHANGES_REQUESTED** [review]
1) The changes look good overall. I'd adjust the outputs in some of the tests though. Not all are needed.
2) Looks like we need to update the `breaking` branch before this can be merged. It still uses Z3 0.8.12 so `b_archlinux` will not pass.
3) It's possible to change the base branch without creating a new PR. You could have just rebased the branch from #12364 on `breaking` and then in the PR clicked "Edit" next to the title - this would show a dropdown where you can select a different base branch.

File: test/cmdlineTests/linking_strict_assembly/args
Comment (mejsiej) [reply test/cmdlineTests/linking_strict_assembly/args:1]: Alright, but this also means I have to change the expected output in the tests.

**Comment by mejsiej** [comment]
I have implemented the changes you have requested @cameel.

File: test/cmdlineTests/yul_function_name_clashes_different_params/args
Comment (cameel) [inline test/cmdlineTests/yul_function_name_clashes_different_params/args]:
```suggestion
--strict-assembly --bin --asm --ir-optimized
```

File: test/cmdlineTests/wasm_to_wasm_memory_instructions_alignment/args
Comment (cameel) [inline test/cmdlineTests/wasm_to_wasm_memory_instructions_alignment/args]:
```suggestion
--yul --yul-dialect ewasm --machine ewasm --bin --ewasm --ewasm-ir --ir-optimized
```

File: test/cmdlineTests/object_compiler/args
Comment (cameel) [inline test/cmdlineTests/object_compiler/args]:
```suggestion
--strict-assembly --optimize --bin --asm --ir-optimized
```

**Review by cameel — APPROVED** [review]
There are 3 more places where args can be removed. Other than that it looks good to merge so I'm approving already.

> Alright, but this also means I have to change the expected output in the tests.

Yeah, that's the whole point. The whole output is very vebose and the new output selection options allow us to limit it to what is relevant in each test.

Note that this is blocked by #12376 though. When @chriseth merges it, you'll need to rebase on `breaking` again so that all CI jobs can pass.

**Comment by mejsiej** [comment]
Alright! Will do it as soon as it gets merged.

**Comment by cameel** [comment]
Oh, one more thing: could you squash your commits? Our convention is to have each commit represent one complete change, without subsequent commits adding stuff and then removing it. Any review fixes should be squashed into the commits they modify before the PR is merged.

**Comment by mejsiej** [comment]
Of course, I will do it when I get home.

**Comment by cameel** [comment]
I have just reviewed #12378 and it reminded me that we need one more thing here: print a message when the output is empty. We currently only have that for compiler mode. Fortunately the situation in assembler mode is simpler than for compilation because we can only have one input file and the file always represents one contract. So we only need a message saying that assembling was successful but no outputs were selected.

**Comment by mejsiej** [comment]
I added the printing of this message.

File: scripts/common_cmdline.sh
Comment (cameel) [inline scripts/common_cmdline.sh]:
```suggestion
        -e '^Assembler run successful, no output requested.$' \
```

**Review by cameel — COMMENTED** [review]
Thanks! Just one small correction to make the regex less susceptible to false-positives.

File: scripts/common_cmdline.sh
Comment (cameel) [reply scripts/common_cmdline.sh]: Actually, I'd add that to all of those excluded lines.

Comment (mejsiej) [reply scripts/common_cmdline.sh]: Got it, added that to every line there.

**Comment by cameel** [comment]
#12376 has been merged today so tests should pass if you rebase on `breaking` now.

**Comment by mejsiej** [comment]
Alright!

**Comment by mejsiej** [comment]
Hey @cameel, how should I do it? When I try to rebase it shows that it will also add 122 commits there and I can't find a way to squash them.

**Comment by cameel** [comment]
That does not sound right. When I try it locally it only tries to rebase the one commit you have here. Is your local `breaking` up to date? And are you sure you have the same `dont-select-any-outputs` you have here checked out and not some older version that was still on `develop`?

Your `breaking` should be at commit 0bbf58ec5e7fc920423e4b4ba4fbfae8612e3af2. The version of `dont-select-any-outputs` that I see is at a3422536dd049df388e05b269b462c878093221d.

**Comment by mejsiej** [comment]
Everything is like you said, this and the breaking branch are exactly like this.

**Comment by cameel** [comment]
But are the local labels you're using at these commits too? They could be out of sync with remote ones. Can you check if the commit ID printed by `git rev-parse HEAD` is what it should be when you check out these branches?

Also, which commands do you use to rebase it? I'm using these:
```bash
git checkout mejsiej/dont-select-any-outputs
git rebase origin/breaking --interactive
```
where:
- `origin` = `git@github.com:ethereum/solidity.git`
- `mejsiej` = `git@github.com:mejsiej/solidity.git`

**Comment by mejsiej** [comment]
Ok, it is now rebased!

**Comment by cameel** [comment]
I think you have pushed the wrong branch. It looks like one of your other PRs.

**Comment by mejsiej** [comment]
No, the problem was that I used the version of CommandLineInterface.cpp from my branch and not breaking. Just changed it right now.

**Comment by mejsiej** [comment]
All of the tests have passed.

**Review by cameel — APPROVED** [review]

**Pull request merged**
